Return type of twisted.internet.defer._ConcurrencyPrimitive.__aexit__ should be Deferred[Literal[False]]

**Describe the incorrect behavior you saw**

I get lots of mypy errors (error: Missing return statement)  when using DeferredLock 

**Describe how to cause this behavior**

Run mypy on:
```
from twisted.internet.defer import DeferredLock

async def foo() -> int:
    lock = DeferredLock()
    async with lock:
        return 5
```

You'll get : `foo.py:3: error: Missing return statement`

Setting the return type to Literal[False] tells mypy that this context manager doesn't swallow exceptions.

**Additional context**
Add any other context about the problem here.

mypy 0.981 (compiled: yes)
Twisted==22.10.0

--- a/src/twisted/internet/defer.py
+++ b/src/twisted/internet/defer.py
@@ -15,7 +15,7 @@ from asyncio import AbstractEventLoop, Future, iscoroutine
 from enum import Enum
 from functools import wraps
 from sys import exc_info
-from types import CoroutineType, GeneratorType, MappingProxyType
+from types import CoroutineType, GeneratorType, MappingProxyType, TracebackType
 from typing import (
     TYPE_CHECKING,
     Any,
@@ -2004,7 +2004,10 @@ class _ConcurrencyPrimitive(ABC, Generic[_DeferredResultT]):
         return self.acquire()
 
     def __aexit__(
-        self, exc_type: bool, exc_val: bool, exc_tb: bool
+        self,
+        __exc_type: Optional[Type[BaseException]],
+        __exc_value: Optional[BaseException],
+        __traceback: Optional[TracebackType]
     ) -> Deferred[Literal[False]]:
         self.release()
         # We return False to indicate that we have not consumed the

--- a/src/twisted/internet/defer.py
+++ b/src/twisted/internet/defer.py
@@ -2003,7 +2003,7 @@ class _ConcurrencyPrimitive(ABC, Generic[_DeferredResultT]):
         """
         return self.acquire()
 
-    def __aexit__(self, exc_type: bool, exc_val: bool, exc_tb: bool) -> Deferred[bool]:
+    def __aexit__(self, exc_type: bool, exc_val: bool, exc_tb: bool) -> Deferred[Literal[False]]:
         self.release()
         # We return False to indicate that we have not consumed the
         # exception, if any.

--- a/src/twisted/internet/defer.py
+++ b/src/twisted/internet/defer.py
@@ -2007,7 +2007,7 @@ class _ConcurrencyPrimitive(ABC, Generic[_DeferredResultT]):
         self,
         __exc_type: Optional[Type[BaseException]],
         __exc_value: Optional[BaseException],
-        __traceback: Optional[TracebackType]
+        __traceback: Optional[TracebackType],
     ) -> Deferred[Literal[False]]:
         self.release()
         # We return False to indicate that we have not consumed the

--- a/src/twisted/internet/defer.py
+++ b/src/twisted/internet/defer.py
@@ -2003,7 +2003,9 @@ class _ConcurrencyPrimitive(ABC, Generic[_DeferredResultT]):
         """
         return self.acquire()
 
-    def __aexit__(self, exc_type: bool, exc_val: bool, exc_tb: bool) -> Deferred[Literal[False]]:
+    def __aexit__(
+        self, exc_type: bool, exc_val: bool, exc_tb: bool
+    ) -> Deferred[Literal[False]]:
         self.release()
         # We return False to indicate that we have not consumed the
         # exception, if any.

--- a/src/twisted/newsfragments/11795.feature
+++ b/src/twisted/newsfragments/11795.feature
@@ -1,0 +1,1 @@
+twisted.internet.defer._ConcurrencyPrimitive.__aexit__ now has a more precise type annotation.

--- a/src/twisted/test/test_defer.py
+++ b/src/twisted/test/test_defer.py
@@ -3518,6 +3518,7 @@ class CoroutineContextVarsTests(unittest.TestCase):
         L{DeferredLock} can be used as an asynchronous context manager.
         """
         lock = DeferredLock()
+        d = None  # This makes mypy check that async with doesn't suppress exceptions.
         async with lock:
             self.assertTrue(lock.locked)
             d = lock.acquire()

